### ISSUE TITLE
restore_part_html() does not correctly matches the tags it removes

### ISSUE THREAD

**Issue by snipfoo** [rstudio/bookdown#1083]
Hi,

It seems there is an issue with the function [`restore_part_html()`](https://github.com/rstudio/bookdown/blob/master/R/html.R#L857) in `R/html.R` as called in [`split_chapters()`](https://github.com/rstudio/bookdown/blob/master/R/html.R#L309) (i.e., with `remove = TRUE`): once the part heading `<h1>` tag is found, the function clears the corresponding line, along with the previous and the next lines (see the `x[i] = x[i - 1] = x[i + 1] = ''` statement on [line 862](https://github.com/rstudio/bookdown/blob/master/R/html.R#L862)). By doing so, the function expects to remove the `<h1>` tag along with the enclosing `<div>` tag.

However, if there is a comment between the part heading and the following chapter heading in the source file (see the example below), the line following the `<h1>` part heading tag will not be the closing `</div>` tag, but the first line of the comment. The call to `restore_part_html()` will then remove this line and leave a dangling `</div>` tag (and, maybe, the rest of the comment, which will end up as regular text in the produced HTML, as its opening tag was removed).

For instance, the following minimal `index.Rmd` will trigger this issue: 
```rmarkdown
---
title: "My book"
date: "`r Sys.Date()`"
---

# My preface {-}

Lorem ipsum dolor sit amet, and so on.

# (PART) My part {-}

<!--
  This comment breaks restore_part_html()
-->

# My chapter

```

The call to `restore_part_html()` will miss the `</div>` tag and remove the opening `<!--` tag instead, resulting in the following invalid HTML as produced by `bookdown::render_book('index.Rmd', 'bookdown::gitbook')`:
```html
<div id="header">
<h1 class="title">My book</h1>
<p class="date"><em>2021-02-21</em></p>
</div>
<div id="my-preface" class="section level1 unnumbered" number="">
<h1>My preface</h1>
<p>Lorem ipsum dolor sit amet, and so on.</p>
</div>



  This comment breaks restore_part_html()
-->
</div>
```

I guess a solution might be to match the actual opening and closing tags of the enclosing `<div>` element, but I'm unfortunately not fluent enough in R to propose a patch fixing this.

By the way, my R environment:
```r
> xfun::session_info('bookdown')
R version 4.0.3 (2020-10-10)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Debian GNU/Linux bullseye/sid

Locale:
  LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
  LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
  LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
  LC_PAPER=en_US.UTF-8       LC_NAME=C                 
  LC_ADDRESS=C               LC_TELEPHONE=C            
  LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

Package version:
  base64enc_0.1.3   bookdown_0.21.6   digest_0.6.27     evaluate_0.14    
  glue_1.4.2        graphics_4.0.3    grDevices_4.0.3   highr_0.8        
  htmltools_0.5.1.1 jsonlite_1.7.2    knitr_1.31        magrittr_2.0.1   
  markdown_1.1      methods_4.0.3     mime_0.10         rlang_0.4.10     
  rmarkdown_2.7.1   stats_4.0.3       stringi_1.5.3     stringr_1.4.0    
  tinytex_0.29      tools_4.0.3       utils_4.0.3       xfun_0.21        
  yaml_2.2.1       
```

Thank you very much in advance!
SnipFoo

---

By filing an issue to this repo, I promise that

- [x] I have fully read the issue guide at https://yihui.org/issue/.
- [x] I have provided the necessary information about my issue.
    - If I'm asking a question, I have already asked it on Stack Overflow or RStudio Community, waited for at least 24 hours, and included a link to my question there.
    - If I'm filing a bug report, I have included a minimal, self-contained, and reproducible example, and have also included `xfun::session_info('bookdown')`. I have upgraded all my packages to their latest versions (e.g., R, RStudio, and R packages), and also tried the development version: `remotes::install_github('rstudio/bookdown')`.
    - If I have posted the same issue elsewhere, I have also mentioned it in this issue.
- [x] I have learned the Github Markdown syntax, and formatted my issue correctly.

I understand that my issue may be closed if I don't fulfill my promises.

**Comment by snipfoo** [comment]
Hi,

Thanks for the quick reply!

I actually encountered this issue by having the part title at the end of a separate file `0_title.Rmd`, followed by chapters in their own standalone files `1_foo.Rmd`, `2_bar.Rmd`, and so on, each chapter file starting with the corresponding chapter title. Even though the part title was on the last line of `0_title.Rmd` and the first chapter title was on the first line of `1_foo.Rmd`, the fact that bookdown inserted extra comments such as `<!--chapter:end:0_title.Rmd-->` at the end of each file before concatenating them triggered the bug.

Maybe that's just a very unusual use case on my part but, before this is fixed, perhaps the doc should also state that the part title should also be in the same file as the first chapter title?

Thanks again!
SnipFoo

**Comment by cderv** [comment]
> perhaps the doc should also state that the part title should also be in the same file as the first chapter title?

Yes currently I believe the PART title should be in the same file. We could precise that in the documentation. Thanks !

**Comment by jtbayly** [comment]
> This issue is part of the improvement to be made regarding parsing and modifying the HTML.

Is this something that is currently underway, @cderv ? Can I track it anywhere?

**Comment by cderv** [comment]
You mean changing the way it currently works ? 

PART title should be in the same file. I did not yet look into how it could evolve. TBH I am not sure it is as simple and it does not feel too much constraint to have a PART header being into the same file as the first title of this part. Is it ? 

I think changing that would require to rewrite the splitting logic of the HTML which may come at some point but not in the very near future.

What is the limitation for you ?

**Comment by jtbayly** [comment]
I don’t use PART currently.  But I’ve raised other issues related to limitations with the current HTML parsing. 

It just sounded to me from your comment like parsing was being reworked. Just a misunderstanding.

**Comment by cderv** [comment]
Don't be sorry! You are completely right to remind us about this not-so-good experience of yours. 

I am aware of your other issues. Reworking the parsing logic for the splitting process is not a light task IMO if we want to do that safe and efficient (like parsing HTML with **xml2**) - I had a quick look and it is not that simple. But it is in my topic to look into more closely but not on the top right now.

Thanks a lot for the reminder. And keep sharing your feedback ! I just wanted to know more if there was more to know about. 

Thanks for your feedback !
